analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs

```diff
@@ -309,4 +313,30 @@ private static ProgramState InitLocals(ControlFlowBranch branch, ProgramState st
         region.Kind == ControlFlowRegionKind.Finally
         || thrown == ExceptionState.UnknownException
         || thrown.Type.DerivesFrom(region.ExceptionType);
+
+    private static IEnumerable<int> DetectLoopBlockOrdinals(ControlFlowGraph cfg)
+    {
+        var reachableFrom = cfg.Blocks.ToImmutableDictionary(x => x.Ordinal, x => new HashSet<int>());
+        var toProcess = new SortedSet<int>(new[] { 0 });
+        while (toProcess.Any())
+        {
+            var current = toProcess.Min;
+            toProcess.Remove(current);
+            var currentReachableFrom = reachableFrom[current];
+            foreach (var successor in cfg.Blocks[current].SuccessorBlocks.Select(x => x.Ordinal))
+            {
+                var successorReachableFrom = reachableFrom[successor];
+                var changed = false;
+                foreach (var block in currentReachableFrom.Append(current))
+                {
+                    changed |= successorReachableFrom.Add(block);
+                }
+                if (changed)
+                {
+                    toProcess.Add(successor);
+                }
+            }
+        }
+        return cfg.Blocks.Select(x => x.Ordinal).Where(x => reachableFrom[x].Contains(x));
```

Comment:
This took me a minute to understand what is going on, with a CFG open on the other screen to run it in my head.
Idea: why not use a stupid DFS for each block?

```csharp
var loopBlocks = new List<int>();
foreach (var block in cfg.Blocks.Select(x => x.Ordinal).ToHashSet())
{
    Dfs(block);
}
return loopBlocks;

void Dfs(int block)
{
    var processed = new HashSet<int>();
    var toProcess = new Stack<int>();
    toProcess.Push(block);
    while (toProcess.Any())
    {
        var current = toProcess.Pop();
        if (processed.Add(current))
        {
            foreach (var successor in cfg.Blocks[current].SuccessorBlocks.Select(x => x.Ordinal))
            {
                if (block == successor)
                {
                    loopBlocks.Add(successor);
                    break;
                }
                else
                {
                    toProcess.Push(successor);
                }
            }
        }
    }
}
```

I haven't really tested it and it might need fine-tuning, but I think it's a bit simpler and you don't need keep as much memory allocated